### ISSUE TITLE
Use locals.scm for highlighting

### ISSUE THREAD

**Issue by Kavantix** [nvim-treesitter/nvim-treesitter#3098]
**Is your feature request related to a problem? Please describe.**
Currently only the highlights.scm is used to determine highlighting of a file. Tree-sitter also has a highlight command and that will also use the locals.scm queries giving much better syntax highlighting with parsers that support it.

**Describe the solution you'd like**
Add support for the locals queries for highlighting

**Describe alternatives you've considered**
I've also opened an issue againt neovim itself neovim/neovim#19181 but was sent here

**Additional context**

**Examples**

With locals:

<img width="229" alt="image" src="https://user-images.githubusercontent.com/6243755/176882094-ac64165f-8855-432a-b0ae-8151a9532818.png">

Without locals:

<img width="228" alt="image" src="https://user-images.githubusercontent.com/6243755/176882215-1765eb4c-8cee-4cad-814d-ce46683d6165.png">

**Docs**

[Here](https://github.com/tree-sitter/tree-sitter/blob/01df16ca9faa1635909ebea242deac200624d9ee/highlight/src/lib.rs#L734-L808) is the implementation of how the tree-sitter cli highlight command handles locals
And the docs of tree sitter highlighting [here](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#local-variables)

**Comment by clason** [comment]
This will likely require changes in Neovim core, but I want to discuss here what the general design should be, as the core here involves adapting captures and highlight groups.

**Comment by clason** [comment]
> The hightlighting in core is not related to this project at all. 

I disagree. This project is very much (also) an out-of-tree rapid prototype for stuff that _will_ be in core. We will probably not bundle all 100+ parsers, for sure, and the first step (possibly even for 0.8?) is to ship the "core" languages in core, but there is all likelihood that we will incrementally add languages as they become sufficiently stable.

And this very much includes highlight groups; it would be very unfortunate if nvim-treesitter and Neovim core used different highlight groups for _treesitter highlighting_. If we need to make a hard fork because we can't agree on a common approach, I would count that as a project failure.

To return to this issue: My point in punting the question here was _exactly_ that this is nvim-treesitter business to decide. _If_ code changes in core were needed, it should be driven by nvim-treesitter. (This is apparently not the case, so this point is moot anyway.)

The technical points about performance are appreciated.

**Comment by Kavantix** [comment]
@theHamsta thanks for explaining why it currently isn't working.
Could you point me to the place where I can enable using the locals for highlighting, so I can test with it?

**Comment by theHamsta** [comment]
@Kavantix use that predicate https://github.com/nvim-treesitter/nvim-treesitter/blob/master/lua/nvim-treesitter/query_predicates.lua#L68 like `(#is? @parameter "parameter")` will check whether the definition of the node is a `@defintion.parameter` in locals for that the captured node needs to be a `@reference` in locals.scm.

> that will be in core

The workflow should be

implementation/playground > documentation > stabilization&feedback > upstream

when we lock this plugin into what is used in nvim, it makes both experimentation and incremental upstreaming difficult as it is more difficult for both to coexist. It does not make sense for us as a playground that nvim is normative, like to document something in core and then to try to implement it here. The benefit of starting with something aribitrary here is that is allowed to be bad. While design decisions might be fun, I think that the person implementing or suggesting something in a PR is a lot wiser than we as they needed to think about that feature. Also most of Neovim user don't participate in design decisions, but they will complain when something doesn't work here for their language. I think it's important to have this experimentation phase here and also try to have it work for the language we maintain here as people will start to get engage when it affects their language. We also had (better?) implementations for folding or for the "locals" highlighting implemented in different plugins than the ones here. Having such competing implementations is also more difficult when there are already changes in core.

I would like to give people implementing features full freedom and have an implementation or a concrete suggestion of a change before we try to find a general solution in core Neovim. For this issue: anybody who has an idea how this can be solved, please go a ahead and do it in the way you think this is solved best!

@steelsojka and me tried to implement this before, but didn't find time/motivation to update and finish it. Someone investing their time to implement this will be my absolute hero ❤️.

**Comment by clason** [comment]
> when we lock this plugin into what is used in nvim, [...]

we seem to be talking past each other -- none of this contradicts what I wrote (to the contrary, it corroborates my explanation why I punted this issue from Neovim to nvim-treesitter)?

**Comment by theHamsta** [comment]
From previous discussion about this feature, there was also the wish to have this feature opt-in (even though it should run much faster in future). An easy intro-PR would be to make the current functionality opt-in in some way (e.g. by ignoring queries with `#is?` during highlighting or move those semantic queries to a separate query category that gets conditionally appended to the normal queries). Then, people could experiment with it and have a motivation to make it faster.

**Comment by Kavantix** [comment]
Wait, does this mean that I have to add a query to the highlights.scm to enable this?

I have to take a look at the code, but that somehow sounds like it will for every token again check if it is a parameter by going to the tree. Which is a completely different approach than the tree sitter cli uses.

**Comment by ahelwer** [comment]
@theHamsta you're looking for someone to finish #1120? What remains?

**Comment by theHamsta** [comment]
@ahelwer an efficient algorithm. #1120 was efficient, since there was a cache in tree-sitter to look up `:parent()`. This got now removed and `:parent()` should be avoided. I don't know whether Atom has now a runtime implementation of highlighting that uses locals. Anything that does not use the helper methods in `lua/nvim-treesitter/query.lua` should be already much faster.

One idea that I had after `:parent()` got removed was to use something like https://en.wikipedia.org/wiki/Interval_tree to a representation of nested scopes that can be efficiently updated for incrementally with the change updates of the language tree https://github.com/theHamsta/neovim/blob/5099653ae1089e9180c560b7a4172674d0fa5398/runtime/lua/vim/treesitter/languagetree.lua#L39-L45 You would need to be able to tell fast in which scope you are currently in and which definitions are in this scope (in case of failure to get the next bigger scope).

My algorithm in #1120:
- building data structure
      - find containing scope of definition by calling `:parent()` on definition until one of the ancestor is found in the hashmap of all scope
      - add definition as a child to that scope
- look up:
     - find containing scope by calling `:parent()` until root is reached
     - if `:parent()` is contained in the hashmap of all scopes, check whether the scope contains a definition with the same name as the reference, if yes, then definition found

This was based on the assumption that `:parent()` is fast, which it was at that moment.

**Comment by theHamsta** [comment]
You can also try how fast or slow #1120 is. Maybe it is worth to just do one complete traversal of the tree to look up parent again efficiently (or better the containing scope directly). So you would basically do the inverse, from the root walk all children and remember when you are currently in a scope, when you are in a scope store the parent scope of all definitions you encounter during your traversal.

**Comment by clason** [comment]
This is by design; even upstream tree-sitter has moved away from using locals queries for highlighting. (This is a job better handled by language servers.)

**Comment by ahelwer** [comment]
Is that so? Interesting. I thought this type of simple semantic highlighting was a big selling point of tree-sitter highlighting compared to regexes.
